### ISSUE TITLE
There seems to exist a bug when using trainer class

### ISSUE THREAD

**Issue by Taise228** [microsoft/Semi-supervised-learning#211]
## Bug

When we train our models using [trainer class](https://github.com/microsoft/Semi-supervised-learning/blob/6cf697e6968f9f77040a9e0162306436a91a00dc/semilearn/lighting/trainer.py#L15), model weights will not be updated during execution. This is because neither `loss.backward()` nor `loss_scaler.step()` are called while training.

As [algorithm.train](https://github.com/microsoft/Semi-supervised-learning/blob/6cf697e6968f9f77040a9e0162306436a91a00dc/semilearn/core/algorithmbase.py#L285) goes, these backward propagation steps should be called via hooks.

### Reproduce the Bug

```{python}
from semilearn import Trainer, get_algorithm, get_data_loader, get_net_builder

config = get_config(...)   # set config to use flexmatch as algorithm
algorithm = get_algorithm(config, get_net_builder(config.net, from_name=False), tb_log=None, logger=None)
train_lb_loader = get_data_loader(...)
train_ulb_loader = get_data_loader(...)
eval_loader = get_data_loader(...)

trainer = Trainer(config, algorithm)
trainer.fit(train_lb_loader, train_ulb_loader, eval_loader)
```

### Error Messages and Logs

Parameters do not change, so every time validation is performed, the same output will be emitted. Same accuracy (both top-1 and top-5), confusion matrix, etc.
